src/internal_modules/roc_core/pool_impl.h

```diff
@@ -30,10 +30,39 @@ namespace core {
 //!
 //! @see Pool.
 class PoolImpl : public NonCopyable<> {
+private:
+    struct Slab : ListNode {};
+    struct Slot : ListNode {};
+
+    // Slot layout when supplying data to user.
+    struct UserSlot {
+        PoolImpl* owner;
+        AlignMax canary_before;
+        AlignMax data[];
+        // Another canary immediately after actual data. Any padding for data is
+        // considered part of canary.
+    };
+
 public:
     //! Size for canary guard.
     enum { CanarySize = sizeof(AlignMax) };
 
+    //! Size of slot given object size.
+    //! Used at compile time to calculate embedded storage in main Pool class.
+    template <size_t ObjectSize> struct SlotSize {
+    private:
+        enum {
+            slot_size_ = MaxAlignedSize<sizeof(Slot)>::value,
+            user_slot_size_ = MaxAlignedSize<sizeof(UserSlot)>::value
+                + MaxAlignedSize<ObjectSize>::value + PoolImpl::CanarySize,
+        };
+
+    public:
+        enum {
+            value = user_slot_size_ > slot_size_ ? user_slot_size_ : slot_size_,
+        };
+    };
```

Comment:
Seems these helpers (along with MaxAlignedSize and AlignMaxSize) are getting out of control and become a bit too complicated. I'd prefer to keep things simple here.

What do you think about the following approach?

In PoolImpl, define two types:

```cpp
class PoolImpl {
public:
    ...

    struct SlotHeader {
        PoolImpl* owner;
        AlignMax data[];
    };

    typedef AlignMax SlotCanary;
};
```

In PoolImpl methods, use SlotHeader struct for ownership check, and compute pointers to SlotCanary (before and after) manually, as you did it before this PR.

Then in Pool class, you can use relatively simple code:

```cpp
enum {
    // header + canary + payalod + canary, maximum-aligned
    SlotSize = (sizeof(PoolImpl::SlotHeader) + sizeof(PoolImpl::SlotCanary)*2 + sizeof(T)
                + sizeof(AlignMax) - 1)
        / sizeof(AlignMax) * sizeof(AlignMax)
};

AlignedStorage<EmbeddedCapacity * SlotSize> embedded_data_;
```

This way, MaxAlignedSize<T>, AlignMaxSize, and SlotSize<T> can be removed.

And also before- and after- canaries will be handled uniformly.

Reply:
Yes. It is looking complicated now. Will go for new approach.